environment_test.go

```diff
@@ -53,6 +53,34 @@ func TestEnvironment_Basic(t *testing.T) {
 	env.chain.assertFailed(t)
 }
 
+type customReporter struct {
+	testing  *testing.T
+	reported bool
+}
+
+func newCustomReporter(t *testing.T) *customReporter {
+	return &customReporter{testing: t}
+}
+
+func (c *customReporter) Errorf(message string, args ...interface{}) {
+	env := NewEnvironment(c)
+	env.Get("key1")
+	c.testing.Logf("Fail: "+message, args...)
+	c.reported = true
+}
+func TestEnvironment_Deadlock(t *testing.T) {
```

Comment:
Let's rename this test to `TestEnvironment_Reentrant`.

TestEnvironment_Deadlock name suggests that we want that env has deadlock.

TestEnvironment_Reentrant name suggests that env is reentrant, i.e. it's ok if it indirectly calls itself (via reporter) and there is *no* deadlock.

https://en.wikipedia.org/wiki/Reentrancy_(computing)